underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -0,0 +1,693 @@
+   * @param dst the destination key to copy to
+   * @return true if the operation was successful, false otherwise
+   */
+  private boolean copy(String src, String dst) {
+    src = stripPrefixIfPresent(src);
+    dst = stripPrefixIfPresent(dst);
+    LOG.debug("Copying {} to {}", src, dst);
+    // Retry copy for a few times, in case some AWS internal errors happened during copy.
+    int retries = 3;
+    for (int i = 0; i < retries; i++) {
+      try {
+        CopyObjectRequest request = new CopyObjectRequest(mBucketName, src, mBucketName, dst);
+        if (Configuration.getBoolean(Constants.UNDERFS_S3A_SERVER_SIDE_ENCRYPTION_ENABLED)) {
+          ObjectMetadata meta = new ObjectMetadata();
+          meta.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+          request.setNewObjectMetadata(meta);
+        }
+        mManager.copy(request).waitForCopyResult();
+        return true;
+      } catch (AmazonClientException | InterruptedException e) {
+        LOG.error("Failed to copy file {} to {}", src, dst, e);
+        if (i != retries - 1) {
+          LOG.error("Retrying copying file {} to {}", src, dst);
+        }
+      }
+    }
+    LOG.error("Failed to copy file {} to {}, after {} retries", src, dst, retries);
+    return false;
+  }
+
+  /**
+   * Internal function to delete a key in S3.
+   *
+   * @param key the key to delete
+   * @return true if successful, false if an exception is thrown
+   */
+  private boolean deleteInternal(String key) {
+    try {
+      if (isFolder(key)) {
+        String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+        mClient.deleteObject(mBucketName, keyAsFolder);
+      } else {
+        mClient.deleteObject(mBucketName, stripPrefixIfPresent(key));
+      }
+    } catch (AmazonClientException e) {
+      LOG.error("Failed to delete {}", key, e);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Gets the child name based on the parent name.
+   *
+   * @param child the key of the child
+   * @param parent the key of the parent
+   * @return the child key with the parent prefix removed, null if the parent prefix is invalid
+   */
+  private String getChildName(String child, String parent) {
+    if (child.startsWith(parent)) {
+      return child.substring(parent.length());
+    }
+    LOG.error("Attempted to get childname with an invalid parent argument. Parent: {} Child: {}",
+        parent, child);
+    return null;
+  }
+
+  /**
+   * @param key the key to get the object details of
+   * @return {@link ObjectMetadata} of the key, or null if the key does not exist
+   */
+  private ObjectMetadata getObjectDetails(String key) {
+    try {
+      if (isFolder(key)) {
+        String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+        return mClient.getObjectMetadata(mBucketName, keyAsFolder);
+      } else {
+        return mClient.getObjectMetadata(mBucketName, stripPrefixIfPresent(key));
```

Comment:
In case key is a file, there are atleast 2 RPCs. Is this the common case?. If yes, the RPC with keyAsFile should be made first. Feel free to ignore if we do not know the answer.